File: BackEnd/JavaScript/javascript-express-npm-prisma-postgresql/src/activity/activity.service.js
import {
  countRecentViews,
  countSearchHistories,
  deleteRecentViews,
  deleteSearchHistories,
  deleteSearchHistoryById,
  findRecentViews,
  findSearchHistories,
} from "./activity.repository.js";
import { notFound } from "../utils/http-error.js";

export async function getRecentViews(userId, query) {
  const page = Math.max(Number(query?.page ?? 1), 1);
  const limit = Math.min(Math.max(Number(query?.limit ?? 20), 1), 100);
  const [items, total] = await Promise.all([
    findRecentViews(userId, page, limit),
    countRecentViews(userId),
  ]);

  return {
    items,
    meta: { total, page, limit },
  };
}

export async function clearRecentViews(userId) {
  await deleteRecentViews(userId);
  return { message: "View history cleared" };
}

export async function getSearchHistories(userId) {
  const [items, total] = await Promise.all([
    findSearchHistories(userId),
    countSearchHistories(userId),
  ]);
  return { items, meta: { total } };
}

export async function clearSearchHistories(userId) {
  await deleteSearchHistories(userId);
  return { message: "Search history cleared" };
}

export async function deleteSearchHistory(userId, id) {
  const result = await deleteSearchHistoryById(userId, id);
  if (result.count === 0) {
    throw notFound("Search history not found");
  }
  return { message: "Search history deleted" };
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/activity/ActivityService.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.activity

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.PbShopException
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.StubResponse
import io.ktor.http.HttpStatusCode

class ActivityService(
    private val repository: ActivityRepository,
) {
    fun views(
        userId: Int,
        page: Int,
        limit: Int,
    ): StubResponse {
        val queryPage = page.coerceAtLeast(1)
        val queryLimit = limit.coerceIn(1, 100)
        val result = repository.listViewHistory(userId, queryPage, queryLimit)
        return StubResponse(
            data =
                result.items.map {
                    mapOf(
                        "id" to it.id,
                        "productId" to it.productId,
                        "productName" to it.productName,
                        "thumbnailUrl" to it.thumbnailUrl,
                        "viewedAt" to it.viewedAt.toString(),
                    )
                },
            meta = pageMeta(queryPage, queryLimit, result.totalCount),
        )
    }

    fun clearViews(userId: Int): StubResponse {
        repository.clearViewHistory(userId)
        return StubResponse(data = mapOf("message" to "View history cleared."))
    }

    fun searches(
        userId: Int,
        page: Int,
        limit: Int,
    ): StubResponse {
        val queryPage = page.coerceAtLeast(1)
        val queryLimit = limit.coerceIn(1, 100)
        val result = repository.listSearchHistories(userId, queryPage, queryLimit)
        return StubResponse(
            data =
                result.items.map {
                    mapOf(
                        "id" to it.id,
                        "keyword" to it.keyword,
                        "createdAt" to it.createdAt.toString(),
                    )
                },
            meta = pageMeta(queryPage, queryLimit, result.totalCount),
        )
    }

    fun clearSearches(userId: Int): StubResponse {
        repository.clearSearchHistories(userId)
        return StubResponse(data = mapOf("message" to "Search history cleared."))
    }

    fun deleteSearch(
        userId: Int,
        searchId: Int,
    ): StubResponse {
        val existing =
            repository.findSearchHistoryById(searchId)
                ?: throw PbShopException(HttpStatusCode.NotFound, "SEARCH_HISTORY_NOT_FOUND", "검색 기록을 찾을 수 없습니다.")
        if (existing.userId != userId) {
            throw PbShopException(HttpStatusCode.Forbidden, "SEARCH_HISTORY_FORBIDDEN", "본인의 검색 기록만 삭제할 수 있습니다.")
        }
        repository.deleteSearchHistory(searchId)
        return StubResponse(data = mapOf("message" to "Search history entry deleted."))
    }

    private fun pageMeta(
        page: Int,
        limit: Int,
        totalCount: Int,
    ): Map<String, Int> =
        mapOf(
            "page" to page,
            "limit" to limit,
            "totalCount" to totalCount,
            "totalPages" to if (totalCount == 0) 0 else ((totalCount + limit - 1) / limit),
        )
}
